### ISSUE TITLE
Walltime for an individual job when running Toil on HPC

### ISSUE THREAD

**Issue by ielis** [DataBiosphere/toil#3037]
Dear developers,
I am using Toil for running a pipeline in a HPC environment that uses Slurm for job scheduling.

Certain jobs of my pipeline are shorter than the others and it would be help me a lot if there was an option for specifying `walltime` for the job. I was not able to find this in the docs. 
Is there a way how to tell the scheduler how long a job is supposed to be running? Any other way than using `$TOIL_SLURM_ARGS`?

Thanks,
Daniel

┆Issue is synchronized with this [Jira Task](https://ucsc-cgl.atlassian.net/browse/TOIL-539)
┆Issue Number: TOIL-539

**Comment by DailyDreaming** [comment]
@ielis Unfortunately, that's a scheduler specific param and there isn't a second way of passing it through at the moment.  We welcome contributions though if this is something you'd like to see.

**Comment by ielis** [comment]
Hi @DailyDreaming , thanks for letting me know what the status is. I might look into that and I'll create a PR if I am able to implement the feature reasonably well.

All the best!

**Comment by mr-c** [comment]
This is needed for a CWL v1.1 feature: https://www.commonwl.org/v1.1/CommandLineTool.html#ToolTimeLimit

**Comment by adamnovak** [comment]
I don't think this would be the right way to implement CWL's `ToolTimeLimit`; that option specifies:

> The execution duration excludes external operations, such as staging of files, pulling a docker image etc, and only counts wall-time for the execution of the command line itself.

Since we do all that staging and pulling inside the Toil job, we can't pass the tool's limit through to the batch system and apply it to the whole Toil job. We'd have to time the tool ourselves after doing the setup.


Kishwar in our lab has been complaining that we should use wall-clock time limits on our Kubernetes cluster; if Toil workflows could carry wall clock time information that would make it easier for us to actually do that.

**Comment by mr-c** [comment]
@adamnovak the pulling of docker containers could be moved outside of the CWL step invocations. There shouldn't be any staging for typical slurm clusters using a shared filesystem.

**Comment by tetron** [comment]
Also you could add some slack time to the time limit to account for job setup.  The intention of that language is to avoid penalizing the user for infrastructure actions that they can't control directly.

**Comment by adamnovak** [comment]
We could definitely pass it through or pass it through plus some slack, as an initial implementation.

Maybe it's time to implement time requests in Toil's backend.

**Comment by lonbar** [comment]
I have recently written an initial implementation of this feature as it would be very useful for me to have it. My humble efforts can be seen [here](https://github.com/DataBiosphere/toil/compare/master...lonbar:toil:issues/featurebranch), but as a quick overview, I have:

* added a `walltime` input to toil which takes an `int` and passes it to slurm's `--time` flag,
* added a conversion function which converts the `walltime` argument (in seconds) to a days-HH:MM:SS format, which is printed for the user,
* exposed this to CWL so that the input of a step's `ToolTimeLimit` is used as `walltime` argument.

There is room for more improvements, I think. If others think this feature is useful I am happy to keep working on this, but would appreciate some guidance on the following:

* This feature is not exposed to WDL. I am not familiar with it and have, so far, been unable to get it to run.
* I did not add a slack time. It seems recent versions of toil pull the docker images ahead of the workflow execution. It might still be useful to compensate for container start-up time, but in this case it is not clear to me what the most appropriate place would be to introduce this slack.
* I have tried to stay close to coding style in the surrounding code, but possibly my additions don't match the current coding style developers. Any suggestions here are very welcome!
* I am having some trouble with some of the tests. Some fail, but these appear in different parts of the codebase seemingly unrelated to my contributions, so I am not sure how to fix them.

**Comment by mr-c** [comment]
@lonbar It looks pretty good to me; can you open a draft PR so we can comment on it?

**Comment by adamnovak** [comment]
It turns out WDL doesn't have a wall clock time limit in the spec as of 1.2, and #5502 implements this for CWL (though see #5533 for a way we could match the spec better), so I am going to close this as done!
